### PR TITLE
[#46] Add support for MongoDB indexes

### PR DESCRIPTION
**Pull request by eddiebrissow**
closes das#46

Added support to:

- MongoDB text index
- MongoDB Compound index
- Query node name starting with
- Added explain to integration tests

Changes to public API:
- New methods
  -  get_atoms_by_field
  - get_atoms_by_index
  - get_node_by_name_starting_with
- Signature changes
  -  create_field_index
- Name change 
  - get_matched_node_name -> get_node_by_name

### CODE REVIEW

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 1]
```diff
@@ -87,31 +94,41 @@ class MongoDBIndex(Index):
     def __init__(self, collection: Collection) -> None:
         self.collection = collection
 
-    def create(self, atom_type: str, field: str, **kwargs) -> Tuple[str, Any]:
+    def create(self, atom_type: str, field: str, fields: Optional[List[str]] = None, index_type: Optional[MongoIndexType] = None,**kwargs) -> Tuple[str, Any]:
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:97]:
```suggestion
    def create(
        self,
        atom_type: str,
        field: str,
        fields: Optional[List[str]] = None,
        index_type: Optional[MongoIndexType] = None,
        **kwargs,
    ) -> Tuple[str, Any]:
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 2]
```diff
@@ -87,31 +94,41 @@ class MongoDBIndex(Index):
...
         conditionals = {}
 
         for key, value in kwargs.items():
             conditionals = {key: {"$eq": value}}
             break  # only one key-value pair
 
-        index_id = f"{atom_type}_{self.generate_index_id(field, conditionals)}"
+        if fields:
+            index_id = f"{atom_type}_{self.generate_index_id(','.join(fields), conditionals)}" + (f"_{index_type.value}" if index_type else "")
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:105]:
```suggestion
            index_id = (
                f"{atom_type}_{self.generate_index_id(','.join(fields), conditionals)}"
                (f"_{index_type.value}" if index_type else "")
            )
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 3]
```diff
@@ -87,31 +94,41 @@ class MongoDBIndex(Index):
...
+        else:
+            index_id = f"{atom_type}_{self.generate_index_id(field, conditionals)}" + (f'_{index_type.value}' if index_type else "")
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:107]:
```suggestion
            index_id = (
                f"{atom_type}_{self.generate_index_id(field, conditionals)}"
                (f'_{index_type.value}' if index_type else "")
            )
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 4]
```diff
@@ -87,31 +94,41 @@ class MongoDBIndex(Index):
...
 
-        index_conditionals = {"name": index_id}
 
+        index_type: MongoIndexType = index_type or (MongoIndexType.COMPOUND if fields is not None else MongoIndexType.FIELD)
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:110]:
```suggestion
        index_type: MongoIndexType = (
            index_type
            or
            (MongoIndexType.COMPOUND if fields is not None else MongoIndexType.FIELD)
        )
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 5]
```diff
@@ -842,7 +876,9 @@ def _retrieve_documents_by_index(
             chunk_size = kwargs.pop('chunk_size', 500)
 
             try:
-                kwargs.update(self._retrieve_custom_index(index_id))
+                # Fallback to previous version
+                conditionals = self._retrieve_custom_index(index_id).get('conditionals', self._retrieve_custom_index(index_id))
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:880]:
```suggestion
                conditionals = self._retrieve_custom_index(index_id).get(
                    'conditionals', self._retrieve_custom_index(index_id)
                )
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 6]
```diff
@@ -87,31 +94,41 @@ class MongoDBIndex(Index):
...
+        index_props = {'index_type': index_type, 'conditionals': conditionals, 'index_name': index_id}
+        index_conditionals = {"name": index_id}
+ 
         if conditionals:
-            index_conditionals["partialFilterExpression"] = conditionals
+            index_conditionals["partialFilterExpression"] = index_props['conditionals']
 
-        index_list = [(field, 1)]  # store the index in ascending order
+        if index_type == MongoIndexType.TEXT:
+            index_list = [((f, 'text') for f in fields)] if fields is not None else [(field, 'text')]
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:118]:
```suggestion
            index_list = (
                [((f, 'text') for f in fields)]
                if fields is not None
                else [(field, 'text')]
            )
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 7]
```diff
@@ -87,31 +94,41 @@ class MongoDBIndex(Index):
...
+        else:
+            index_list = [(f, ASCENDING) for f in fields] if fields is not None else [(field, ASCENDING)] # store the index in ascending order
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:120]:
```suggestion
            index_list = (
                [(f, ASCENDING) for f in fields]
                if fields is not None
                else [(field, ASCENDING)] # store the index in ascending order
            )
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 8]
```diff
@@ -912,14 +954,16 @@ def create_field_index(
...
         index_id = ""
 
+        mongo_index_type = MongoIndexType.TEXT if index_type == FieldIndexType.TEXT else None
+
         try:
             exc = ""
-            index_id, conditionals = MongoDBIndex(collection).create(atom_type, field, **kwargs)
-            serialized_conditionals = pickle.dumps(conditionals)
-            serialized_conditionals_str = base64.b64encode(serialized_conditionals).decode('utf-8')
+            index_id, index_props = MongoDBIndex(collection).create(atom_type, field, index_type=mongo_index_type, fields=fields, **kwargs)
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:961]:
```suggestion
            index_id, index_props = MongoDBIndex(collection).create(
                atom_type, field, index_type=mongo_index_type, fields=fields, **kwargs
            )
```

**Review by angeloprobst — CHANGES_REQUESTED** [review]
Just some comments regarding code style.
The rationale is to avoid too long lines.

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 9]
```diff
@@ -353,12 +373,26 @@ def get_matched_node_name(self, node_type: str, substring: str) -> str:
...
             FieldNames.COMPOSITE_TYPE_HASH: node_type_hash,
-            FieldNames.NODE_NAME: {'$regex': substring},
+            '$text': {'$search': substring},
         }
         return [
             document[FieldNames.ID_HASH]
             for document in self.mongo_atoms_collection.find(mongo_filter)
         ]
+    
+    def get_node_starting_with(self, node_type: str, startswith: str):
```
Comment (andre-senna) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:383]: Thius method should be in the superclass and implemented somehow in each other subclass

File: hyperon_das_atomdb/database.py [hunk 10]
```diff
@@ -29,6 +29,11 @@ class FieldNames(str, Enum):
     IS_TOPLEVEL = 'is_toplevel'
 
 
+class FieldIndexType(str, Enum):
+    DEFAULT = 'default'
```
Comment (andre-senna) [inline hyperon_das_atomdb/database.py:33]: DEFAULT is a bad name as it doesn't give any clue about what the index is. Suggestion: "BINARY_TREE"

File: hyperon_das_atomdb/database.py [hunk 11]
```diff
@@ -29,6 +29,11 @@ class FieldNames(str, Enum):
...
+    TEXT = 'text'
```
Comment (andre-senna) [inline hyperon_das_atomdb/database.py:34]: The same here. Suggestion: TOKEN_INVERTED_LIST

**Review by andre-senna — CHANGES_REQUESTED** [review]

**Review by marcocapozzoli — APPROVED** [review]

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 12]
```diff
@@ -87,31 +94,68 @@ class MongoDBIndex(Index):
...
+        field: str,
+        fields: Optional[List[str]] = None,
+        index_type: Optional[MongoIndexType] = None,
+        **kwargs,
+    ) -> Tuple[str, Any]:
         conditionals = {}
 
         for key, value in kwargs.items():
             conditionals = {key: {"$eq": value}}
             break  # only one key-value pair
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:109]: I know this was here previously to your change, but I think it's worth it to refactor it.

```suggestion
        key, value = next(iter(kwargs.items())):
        conditionals = {key: {"$eq": value}}
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 13]
```diff
@@ -87,31 +94,68 @@ class MongoDBIndex(Index):
...
         index_conditionals = {"name": index_id}
-
+ 
         if conditionals:
-            index_conditionals["partialFilterExpression"] = conditionals
+            index_conditionals["partialFilterExpression"] = index_props['conditionals']
+
+        if index_type == MongoIndexType.TEXT:
+            index_list = (
+                [((f, 'text') for f in fields)]
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:137]: It seems that there is an extra pair of `(` and `)`.

```suggestion
                [(f, 'text') for f in fields]
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 14]
```diff
@@ -359,7 +406,59 @@ def get_matched_node_name(self, node_type: str, substring: str) -> str:
...
+            self, 
+            index_id: str, 
+            query: List[OrderedDict[str, str]],
+            cursor: Optional[int] = 0, 
+            chunk_size: Optional[int] = 500
+            ) -> Tuple[int, List[str]]:
+        mongo_filter = OrderedDict([(q['field'], q['value']) for q in query])
+        return self._get_atoms_by_index(index_id, cursor=cursor, chunk_size=chunk_size, **mongo_filter)
+
+    def get_atoms_by_text_field(self, text_value: str, field: Optional[str] = None, text_index_id: Optional[str] = None) -> List[str]:
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:427]:
```suggestion
    def get_atoms_by_text_field(
        self,
        text_value: str,
        field: Optional[str] = None,
        text_index_id: Optional[str] = None,
    ) -> List[str]:
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 15]
```diff
@@ -912,14 +1020,18 @@ def create_field_index(
 
         index_id = ""
 
+        mongo_index_type = MongoIndexType.TEXT if index_type == FieldIndexType.TOKEN_INVERTED_LIST else None
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:1023]:
```suggestion
        mongo_index_type = (
            MongoIndexType.TEXT
            if index_type == FieldIndexType.TOKEN_INVERTED_LIST
            else None
        )
```

File: hyperon_das_atomdb/database.py [hunk 16]
```diff
@@ -272,6 +277,75 @@ def get_matched_node_name(self, node_type: str, substring: str) -> str:
...
+            index_id (str): Index ID created by the function 'create_field_index'
+            query (List[Dict[str, str]]): List of dicts containing 'field' and 'value' keys
+
+        Returns:
+            List[str]: List of node IDs
+        """        
+        ...  # pragma no cover
+
+    @abstractmethod
+    def get_atoms_by_text_field(self, text_value: str, field: Optional[str] = None, text_index_id: Optional[str] = None) -> List[str]:
```
Comment (angeloprobst) [inline hyperon_das_atomdb/database.py:316]:
```suggestion
    def get_atoms_by_text_field(
        self,
        text_value: str,
        field: Optional[str] = None,
        text_index_id: Optional[str] = None,
    ) -> List[str]:
```

File: hyperon_das_atomdb/database.py [hunk 17]
```diff
@@ -272,6 +277,75 @@ def get_matched_node_name(self, node_type: str, substring: str) -> str:
...
+        """
+        Query the database by a text field, use the text_value arg to query using a existing text index (text_index_id is optional),
+        if a TOKEN_INVERTED_LIST type of index wasn't previously created the field arg must be provided or it will raise an Exception.
+        When 'text_value' and 'field' value are provided, it will defaults to a regex search, creating a index to the field can improve
+        the performance.
+
+        Args:
+            text_value (str): Value to search for, if only this argument is provided it will use a TOKEN_INVERTED_LIST index in the search
+            field (Optional[str]): Field to be used to search, if this argument is provided it will not use TOKEN_INVERTED_LIST in the search
+            text_index_id (Optional[str]): TOKEN_INVERTED_LIST index id to search for
```
Comment (angeloprobst) [inline hyperon_das_atomdb/database.py:326]: Could you please make this block comment horizontally shorter?

**Review by angeloprobst — CHANGES_REQUESTED** [review]

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 18]
```diff
@@ -87,31 +94,68 @@ class MongoDBIndex(Index):
     def __init__(self, collection: Collection) -> None:
         self.collection = collection
 
-    def create(self, atom_type: str, field: str, **kwargs) -> Tuple[str, Any]:
+    def create(
+        self,
+        atom_type: str,
+        field: str,
```
Comment (andre-senna) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:100]: Will be fixed in https://github.com/singnet/das-atom-db/issues/153

File: hyperon_das_atomdb/adapters/redis_mongo_db.py (hunk 12 above)
Comment (eddiebrissow) [reply hyperon_das_atomdb/adapters/redis_mongo_db.py:109]: I'll add an IF because kwargs can be empty

**Pull request merged**